setuptools/installer.py

```diff
@@ -37,25 +35,32 @@ def fetch_build_egg(dist, req):
     return _fetch_build_egg_no_warn(dist, req)
 
 
-def _fetch_build_eggs(dist, requires: _StrOrIter) -> list[Distribution]:
-    import pkg_resources  # Delay import to avoid unnecessary side-effects
-
+def _fetch_build_eggs(dist, requires: _StrOrIter) -> list[metadata.Distribution]:
     _DeprecatedInstaller.emit(stacklevel=3)
     _warn_wheel_not_available(dist)
 
-    resolved_dists = pkg_resources.working_set.resolve(
-        _reqs.parse(requires, pkg_resources.Requirement),  # required for compatibility
-        installer=partial(_fetch_build_egg_no_warn, dist),  # avoid warning twice
-        replace_conflicting=True,
+    from . import _reqs
+
+    needed_reqs = (
+        req for req in _reqs.parse(requires) if not req.marker or req.marker.evaluate()
     )
+    resolved_dists = [_fetch_build_egg_no_warn(dist, req) for req in needed_reqs]
     for dist in resolved_dists:
-        pkg_resources.working_set.add(dist, replace=True)
+        # dist.locate_file('') is the directory containing EGG-INFO, where the importabl
+        # contents can be found.
+        sys.path.insert(0, str(dist.locate_file('')))
```

Comment:
The biggest diff is in this change, which instead of relying of the pkg_resources.working_set, simply makes the package available on sys.path. There are probably some scenarios in which `working_set` provides more sophistication (possibly overriding already imported packages), but my instinct is that risk is low given the limited scope of setup_requires outside of pep517 these days.